mm2src/coins/utxo/utxo_common.rs

```diff
@@ -1856,36 +1874,61 @@ where
-        ActualTxFee::Dynamic(fee) => fee,
-    };
+        ActualTxFee::Dynamic(fee) => {
+            // take into account that the dynamic tx fee may increase during the swap
+            let dynamic_fee = coin.increase_dynamic_fee_by_stage(fee, stage);
+
+            let outputs_count = outputs.len();
+            let (unspents, _recently_sent_txs) = try_map!(
+                coin.list_unspent_ordered(&coin.as_ref().my_address).await,
+                TradePreimageError::Other
+            );
 
-    // take into account that the dynamic tx fee may increase during the swap
-    let dynamic_fee = coin.increase_dynamic_fee_by_stage(dynamic_fee, stage);
+            let actual_tx_fee = Some(ActualTxFee::Dynamic(dynamic_fee));
+            let (tx, data) = generate_transaction(coin, unspents, outputs, fee_policy, actual_tx_fee, gas_fee).await?;
 
-    let outputs_count = outputs.len();
-    let (unspents, _recently_sent_txs) = try_map!(
-        coin.list_unspent_ordered(&coin.as_ref().my_address).await,
-        TradePreimageError::Other
-    );
+            let total_fee = if tx.outputs.len() == outputs_count {
+                // take into account the change output
+                data.fee_amount + (dynamic_fee * P2PKH_OUTPUT_LEN) / KILO_BYTE
+            } else {
+                // the change outputs is included already
+                data.fee_amount
+            };
 
-    let actual_tx_fee = Some(ActualTxFee::Dynamic(dynamic_fee));
-    let (tx, data) = generate_transaction(coin, unspents, outputs, fee_policy, actual_tx_fee, gas_fee).await?;
+            Ok(big_decimal_from_sat(total_fee as i64, decimals))
+        },
+        ActualTxFee::FixedPerKb(fee) => {
+            let outputs_count = outputs.len();
+            let (unspents, _recently_sent_txs) = try_map!(
+                coin.list_unspent_ordered(&coin.as_ref().my_address).await,
+                TradePreimageError::Other
+            );
 
-    let total_fee = if tx.outputs.len() == outputs_count {
-        // take into account the change output
-        data.fee_amount + (dynamic_fee * P2PKH_OUTPUT_LEN) / KILO_BYTE
-    } else {
-        // the change outputs is included already
-        data.fee_amount
-    };
+            let (tx, data) = generate_transaction(coin, unspents, outputs, fee_policy, Some(tx_fee), gas_fee).await?;
 
-    Ok(big_decimal_from_sat(total_fee as i64, decimals))
+            let total_fee = if tx.outputs.len() == outputs_count {
+                // take into account the change output if tx_size_kb(tx with change) > tx_size_kb(tx without change)
+                let tx = UtxoTx::from(tx);
+                let tx_bytes = serialize(&tx);
+                if tx_bytes.len() as u64 % KILO_BYTE + P2PKH_OUTPUT_LEN > KILO_BYTE {
+                    data.fee_amount + fee
+                } else {
+                    data.fee_amount
+                }
+            } else {
+                // the change outputs is included already
```

Comment:
I was mistaken when writing this comment: `the change output is included already`.
Could you please correct it?

Reply:
Sure, done :slightly_smiling_face: